applications/serial_lte_modem/src/slm_ppp.c

```diff
@@ -0,0 +1,311 @@
+		return ret;
+	}
+
+	/* Set the PPP MTU to that of the LTE link. */
+	mtu = MIN(mtu, sizeof(ppp_data_buf));
+	net_if_set_mtu(ppp_iface, mtu);
+	LOG_DBG("MTU set to %u.", mtu);
+
+	ret = net_if_up(ppp_iface);
+	if (ret) {
+		LOG_ERR("Failed to bring PPP interface up (%d).", ret);
+		return ret;
+	}
+
+	if (!open_ppp_sockets()) {
+		close_ppp_sockets();
+		net_if_down(ppp_iface);
+		return -ENOTCONN;
+	}
+
+	LOG_INF("PPP started.");
+
+	k_thread_create(&ppp_data_passing_thread_id, ppp_data_passing_thread_stack,
+			K_THREAD_STACK_SIZEOF(ppp_data_passing_thread_stack),
+			ppp_data_passing_thread, NULL, NULL, NULL,
+			K_PRIO_COOP(10), 0, K_NO_WAIT);
+	k_thread_name_set(&ppp_data_passing_thread_id, "ppp_data_passing");
+
+	return 0;
+}
+
+/** @return Whether PPP is fully started. */
+static bool slm_ppp_is_started(void)
+{
+	return net_if_flag_is_set(ppp_iface, NET_IF_UP)
+	    && ppp_fds[ZEPHYR_FD_IDX] >= 0
+	    && ppp_fds[MODEM_FD_IDX] >= 0;
+}
+
+static void slm_ppp_stop(void)
+{
+	if (!slm_ppp_is_started()) {
+		return;
+	}
+	/* First bring the interface down so that slm_ppp_is_started()
+	 * returns false right away. This is to prevent trying to stop PPP at the
+	 * same time from multiple sources: the original one (e.g. "AT#XPPP=0")
+	 * and the data passing thread. The latter attempts to stop PPP when it receives
+	 * an error on some of the sockets, which happens when they are closed.
+	 */
+	const int ret = net_if_down(ppp_iface);
+
+	if (ret) {
+		LOG_WRN("Failed to bring PPP interface down (%d).", ret);
+	}
+
+	close_ppp_sockets();
+
+	k_thread_join(&ppp_data_passing_thread_id, K_SECONDS(1));
+
+	LOG_INF("PPP stopped.");
+}
+
+int slm_ppp_init(void)
+{
+	if (!device_is_ready(ppp_uart_dev)) {
+		return -EAGAIN;
+	}
+
+	ppp_iface_idx = net_if_get_by_name(CONFIG_NET_PPP_DRV_NAME);
+	ppp_iface = net_if_get_by_index(ppp_iface_idx);
+	if (!ppp_iface) {
+		LOG_ERR("PPP network interface not found (%d).", ppp_iface_idx);
+		return -ENODEV;
+	}
+	net_if_flag_set(ppp_iface, NET_IF_POINTOPOINT);
+
+	LOG_DBG("PPP initialized.");
+	return 0;
+}
+
+/* Handles AT#XPPP commands. */
+int handle_at_ppp(enum at_cmd_type cmd_type)
+{
+	if (cmd_type != AT_CMD_TYPE_SET_COMMAND
+	 || at_params_valid_count_get(&slm_at_param_list) != 2) {
+		return -EINVAL;
+	}
+	int ret;
+	unsigned int op;
+	enum {
+		OP_STOP,
+		OP_RESTART,
+		OP_COUNT
+	};
+
+	ret = at_params_unsigned_int_get(&slm_at_param_list, 1, &op);
+	if (ret) {
+		return ret;
+	} else if (op >= OP_COUNT) {
+		return -ENOTSUP;
+	}
+
+	slm_ppp_stop();
+	if (op == OP_RESTART) {
+		return slm_ppp_start();
+	}
+	assert(op == OP_STOP);
+	return 0;
+}
+
+static void ppp_data_passing_thread(void*, void*, void*)
+{
+	const size_t mtu = net_if_get_mtu(ppp_iface);
+	struct pollfd fds[PPP_FDS_COUNT];
+
+	for (size_t i = 0; i != ARRAY_SIZE(fds); ++i) {
+		fds[i].fd = ppp_fds[i];
+		fds[i].events = POLLIN;
+	}
+
+	while (true) {
+		const int poll_ret = poll(fds, ARRAY_SIZE(fds), 1000);
+
+		if (poll_ret == 0) {
```

Comment:
If no action is taken in case of no events, maybe just set the poll timeout to `-1` (infinite)?

Reply:
True.